File: solution/5719/거의 최단 경로.py
f = open('./solution/5719/input.txt', 'r')
input = f.readline

# from sys import stdin
from heapq import heappop, heappush
# input = stdin.readline

MAX = float('inf')

class Node:
    def __init__(self):
        self.dist = MAX;
        self.parent = []
        self.adj = {}
            
def dijkstra(N, nodes, S):
    for i in range(N):
        nodes[i].dist = MAX if i != S else 0
    
    hq = [(0, S)]
    visit = [False for _ in range(N)]
    while hq:
        _, cur = heappop(hq)
        visit[cur] = True
        for to, cost in nodes[cur].adj.items():
            if not visit[to]:
                if nodes[to].dist > nodes[cur].dist + cost:
                    nodes[to].dist = nodes[cur].dist + cost
                    nodes[to].parent = [cur]
                elif nodes[to].dist == nodes[cur].dist + cost:
                    nodes[to].parent.append(cur)
                heappush(hq, (cost, to))

def remove(target, S, nodes):
    if target!=S:
        for p in nodes[target].parent:
            del nodes[p].adj[target]
            remove(p, S, nodes)
    
N, M = map(int, input().split())
while N!=0 and M!=0:
    nodes = [Node() for _ in range(N)]
    S, D = map(int, input().split())
    for _ in range(M):
        u, v, p = map(int, input().split())
        nodes[u].adj[v]=p
    
    dijkstra(N, nodes, S)
    remove(D, S, nodes)
    dijkstra(N, nodes, S)
    
    print(nodes[D].dist if nodes[D].dist!=MAX else -1)
    N, M = map(int, input().split())

File: solution/5719/거의 최단 경로.java
import java.io.*;
import java.util.*;

public class Main {

    static Map<Integer, Integer>[] edges;
    static ArrayList<Integer>[] parent;
    static int N, M, S, D;
    static int[] dist;
    static boolean[][] visit;
    static final int MAX = Integer.MAX_VALUE;

    public static void main(String[] args) throws Exception {
//        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream("./input.txt")));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        while(N!=0 && M!=0){

            st = new StringTokenizer(br.readLine());
            S = Integer.parseInt(st.nextToken());
            D = Integer.parseInt(st.nextToken());

            edges = new HashMap[N];
            for (int i = 0; i < N; i++)
                edges[i] = new HashMap<>();

            for (int i = 0; i < M; i++) {
                st = new StringTokenizer(br.readLine());
                int U = Integer.parseInt(st.nextToken());
                int V = Integer.parseInt(st.nextToken());
                int P = Integer.parseInt(st.nextToken());
                edges[U].put(V, P);
            }

            visit = new boolean[N+1][N+1];
            dijkstra(); //최단경로 찾기
            remove(D); // 최단경로 지우기
            dijkstra(); //최단거리 찾기

            bw.write(dist[D]!=MAX ? dist[D]+"\n" : "-1\n");

            st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            M = Integer.parseInt(st.nextToken());
        }

        bw.flush();
        bw.close();
        br.close();
    }
    static void dijkstra(){
        dist = new int[N];
        parent = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            dist[i] = MAX;
            parent[i] = new ArrayList<>();
        }
        dist[S] = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b)->a[1]-b[1]);
        pq.add(new int[]{S, 0});
        while(!pq.isEmpty()){
            int[] tmp = pq.poll();
            int now = tmp[0];
            if (tmp[1] > dist[now])
                continue;

            for( Map.Entry<Integer, Integer> edge : edges[now].entrySet()){
                int next = edge.getKey();
                int cost = edge.getValue();

                if (dist[now] + cost < dist[next]){
                    dist[next] = dist[now] + cost;
                    pq.add(new int[]{next, dist[next]});
                    parent[next].clear();
                    parent[next].add(now);
                } else if (dist[now] + cost == dist[next])
                    parent[next].add(now);
            }
        }
    }
    static void remove(int child){
        if (child!=S)
            for(int p : parent[child]){
                if (!visit[child][p]) {
                    visit[child][p] = true;
                    edges[p].remove(child);
                    remove(p);
                }
            }
    }
}